core/controllers/skill_editor_test.py

```diff
@@ -52,41 +79,75 @@ def test_access_skill_editor_page(self):
             # Check that non-admins cannot access the editor page.
             self.login(self.NEW_USER_EMAIL)
             response = self.testapp.get(
-                '%s/%s' % (
-                    feconf.SKILL_EDITOR_URL_PREFIX, self.skill_id),
-                expect_errors=True)
+                self.url, expect_errors=True)
             self.assertEqual(response.status_int, 401)
             self.logout()
 
             # Check that admins can access and edit in the editor page.
             self.login(self.ADMIN_EMAIL)
-            response = self.testapp.get(
-                '%s/%s' % (feconf.SKILL_EDITOR_URL_PREFIX, self.skill_id))
+            response = self.testapp.get(self.url)
             self.assertEqual(response.status_int, 200)
             self.logout()
 
-    def test_editable_skill_handler_get(self):
-        # Check that non-admins cannot access the editable skill data.
-        with self.swap(feconf, 'ENABLE_NEW_STRUCTURES', True):
-            self.login(self.NEW_USER_EMAIL)
-            response = self.testapp.get(
-                '%s/%s' % (
-                    feconf.SKILL_EDITOR_DATA_URL_PREFIX, self.skill_id),
-                expect_errors=True)
-            self.assertEqual(response.status_int, 401)
-            self.logout()
+    def test_skill_editor_page_fails(self):
+        self.login(self.ADMIN_EMAIL)
+        # Check GET returns 404 when new strutures' pages are not enabled.
+        with self.swap(feconf, 'ENABLE_NEW_STRUCTURES', False):
+            response = self.testapp.get(self.url, expect_errors=True)
+            self.assertEqual(response.status_int, 404)
 
-            # Check that admins can access the editable skill data.
+        # Check GET returns 404 when cannot get skill by id.
```

Comment:
Typically, the controller tests are integration tests for the server (of a sort). So, instead of mocking here, maybe just do a GET for a skill ID which does not exist?

Reply:
I just tried that. The test passed but the line below is not covered:

https://github.com/oppia/oppia/blob/8f3d11c1ff84f01a02ec0871382bac3895009ad0/core/controllers/skill_editor.py#L66

I use `skill_services.get_new_skill_id()` to get a new skill ID. I don't call `save_new_skill` so `get_skill_by_id` should return `None`. However, it seems like the code never reaches the call to `get_skill_by_id` in this case. The reason I think so is because change the expected error to 500 so that I can add some debugging lines at several places in the code, e.g. in the body of the `get` method. None of those debugging lines. In addition, I got this error message:

```
WARNING:root:Invalid URL requested: http://localhost/skill_editor/F4ynhfByGQvn
```

This skill ID should pass the check of `require_valid_skill_id`, so I'm suspecting somewhere in the code we're checking if the skill exists even before the method handling the GET request is called.

Reply:
Ah! I think you might have found even more dead code, which is great since it means we can remove it.

There's a decorator on the handler which checks the skill beforehand for validity:

https://github.com/oppia/oppia/blob/develop/core/controllers/skill_editor.py#L55

I think that might have a check for validity, so that this one is redundant?

Reply:
So, do want to keep the mock now to make the coverage 100%?

Reply:
I suggest removing the dead code in the other test because the decorator calls `get_skill_rights` so the check `if skill_rights is None` is already handled by the decorator.

In this thread we're discussing tests in which I mock the call to `skill_services.get_skill_by_id`.

In theory, if a skill id doesn't exist, both `get_skill_rights` (called by the decorator) and `get_skill_by_id` (call by the methods in `skill_editor.py`) will return `None`.

In practice, I don't know if the statement above will always hold true, or there might be cases in which the call in the decorator doesn't return `None` but the calls in `skill_editor.py` do, so I would prefer to keep these lines of code: 

```
if skill_id is None:
    raise Exception ...
```

in `skill_editor.py`.

If we want to keep the lines of code above and want to have 100% coverage, we would have to mock `get_skill_by_id`.

So, my question is which option do we choose:
1/ Remove the `if skill_id is None` checks wherever in the code that `get_skill_by_id` is used (dead code).
2/ Keep the `if skill_id is None` checks and mock `get_skill_by_id` in the test to make the coverage 100%

@seanlip ?

Reply:
Ah, so, I think I get it. You're thinking about the case where skill rights is not None, but the skill ID retrieval fails, right? Generally that should not happen, but theoretically it is a possibility since they are two different models.

So, how about doing the reverse of what I suggested for another test? I.e. create the skill the usual way (which causes both a skill model and a skill rights model to be created), then delete the skill model manually. Then you don't need to mock things, and you can still test the line you're referring to.

Reply:
Done